lambdas/preview/index.py

```diff
@@ -141,35 +145,102 @@ def extract_vcf(file_):
     """
     Pull summary info from VCF: meta-information, header line, and data lines
     (in that order, up to MAX_LINES)
-    """
 
+    Args:
+        file_ - file-like
+    """
     meta = []
     header = []
     data = []
-
+    size = 0
     with open(file_.name, 'rb') as vcf:
         for index, line in enumerate(vcf, start=1):
-            # don't escape quotes
-            line = line.strip().decode('unicode_escape')
-            formatted = f'<p>{line}</p>'
+            line = _truncate(line.rstrip().decode(), size)
             if line.startswith('##'):
-                meta.append(formatted)
+                meta.append(line)
             elif line.startswith('#'):
-                header.append(formatted)
+                header.append(line)
             else:
-                data.append(formatted)
+                data.append(line)
+            size += len(line)
+            # stop if we're over the max
+            if index >= MAX_LINES or size > MAX_BYTES:
+                break
 
-            if index >= MAX_LINES:
+    info = {
+        'data': {
+            'meta': meta,
+            'header': header,
+            'data': data,
+        }
+    }
+
+    return '', info
+
+def extract_txt(file_):
+    """
+    Display first and last few lines of a potentially large file.
+
+    Args:
+        file_ - file-like object
+    """
+    size = 0
+    ellipsis = False
+    head = []
+    with open(file_.name, 'rb') as txt:
+        for index, line in enumerate(txt, start=1):
+            line = _truncate(line.rstrip().decode(), size)
+            size += len(line)
+            # this is a heuristic; can fail to return shorter tail lines
+            # if head lines are uncharacteristically long
+            # we're guarding against, for example, huge single-line JSON files
+            if index <= MAX_LINES and size <= MAX_BYTES:
+                head.append(line)
+            if size > MAX_BYTES:
                 break
+            if index > MAX_LINES:
+                ellipsis = True
+                break
+    tail = []
+    if ellipsis: # in this case we need a tail
+        headlen = int(MAX_LINES/2) # pylint: disable=old-division
+        taillen = MAX_LINES - headlen
+        # cut head
+        head = head[:headlen]
+        # grow tail
+        with open(file_.name, 'rb') as txt:
+            count = 0
+            txt.seek(0, os.SEEK_END) # the very end
+            while count < taillen:
+                txt.seek(-2, os.SEEK_CUR)
```

Comment:
This seems crazy inefficient - seek and read, one byte at a time?

Reply:
is there a faster way to find the lines at the end of a file when we don't know line width?

Reply:
oh - we could read it into memory in large chunks, sure

Reply:
Ideally, read from the end in chunks of a few KBs.

Don't go too crazy optimizing this, though: even reading every single line of the file wouldn't be that big of a deal, given that we just downloaded the whole file from S3. But making two syscalls per byte could end up pretty expensive if you have a single really long line.

Technically, though, even reading a single line isn't safe: theoretically, you could have a GB file with no line breaks.

Reply:
we know how many bytes are left in the budget so i can just seek back that far and get as many whole lines as are available

Reply:
Oh yeah, that works